### PR TITLE
Feat/data table/expanded group keys

### PR DESCRIPTION
**Pull request by SergiiTykhenkoMCC**
#### What does this PR do?
Added the ability to set expanded groups by default via expandedGroupKeys prop.
Added onToggle event handler that passes groupState
Resolves issue [#3302](https://github.com/grommet/grommet/issues/3302)

#### Where should the reviewer start?
DataTable/index.d.ts

#### What testing has been done on this PR?
Tested manually.
2 tests where added to DataTable-test

#### How should this be manually tested?
use the storybook

#### What are the relevant issues?
[#3302](https://github.com/grommet/grommet/issues/3302)

#### Do the grommet docs need to be updated?
yes

#### Should this PR be mentioned in the release notes?
yes

#### Is this change backwards compatible or is it a breaking change?
non-breaking

### CODE REVIEW

**Comment by SergiiTykhenkoMCC** [comment]
> Hi buddy, thanks for working on this, regarding:
> 
> > Do the grommet docs need to be updated?
> > yes
> 
> will you be willing to update doc.js and run `yarn test update` to include the new snapshots?
> 
> In addition, I'd love to see a storybook example that demonstrates this behavior and to see if it is compliant with accessibility standards.

Hi Shimi,
yes of course, I was intended to do so but I just hope to hear some respond regarding code changes at first. Weather they're acceptable or not.

**Comment by ericsoderberghp** [comment]
A couple of thoughts on this ...

My initial reaction was to think that `expandedGroupKeys` would always drive the expanded state, and that the caller would use that in conjunction with `onToggle` to manage that state. But, the implementation only uses `expandedGroupKeys` to drive the initial expansion state.

What is the motivation behind wanting `onToggle`? Since the user can't use it to drive subsequent state, I'm not sure what it would be used for. I'm not saying there isn't a reason. I'd just like to know what it is.

The fact that `onToggle` always sends the expanded state of all groups means the caller would have to persist that state and then compare it on subsequent calls to know what changed. In general, when there's a relationship between a `value` and `onChange`, we like to keep the two aligned such that the caller can pass the argument to `onChange` back as the `value`.

I think we should associate the terminology more closely so that it's clearer that `onToggle` is related to `groupBy`. One idea I'm contemplating is whether to do this:
```
groupBy={{ property: 'a', expand: ['one'], onExpand: func(array of values) }}
```
Where `onChangeExpand` would be passed an array of the values that are currently expanded, such that the caller could feed this back into `expand` and that would always control the expansion state. If we didn't wrap everything inside `groupBy`, perhaps `expandGroupBy` and `onExpandGroupBy` would make the naming more aligned?

In the future, making the doc.js changes, with their descriptions first, would be a good way to have a conversation around an interface change first. Not required, but helpful.

**Comment by SergiiTykhenkoMCC** [comment]
Hi ericsoderberghp,
1) Yes, 'expandedGroupKeys' sets just initial expansion state. It was done this way in order not to make component controlled as it will require additional changes.
2) With 'onToggle' we can get current expansion state and use it for other needs, not setting expansion state from it again.
3) I'm good with grouping all props inside groupBy.
Let's decide whether we should go with fully controlled expansion state via 'expandedGroupKeys'(or how it's called in future) or it will just set initial state?

**Comment by ericsoderberghp** [comment]
Can you elaborate an example of "other needs" you mention with regard to `onToggle`?

**Comment by SergiiTykhenkoMCC** [comment]
> Can you elaborate an example of "other needs" you mention with regard to `onToggle`?

For example currently we need to save this state and apply it later, when user will come back to this table. This way we can retrieve expansion state from 'onToggle' action and save it

**Comment by SergiiTykhenkoMCC** [comment]
Hello ericsoderberghp,
I've implemented some additional changes and also created story and updated doc.js

File: src/js/components/DataTable/buildState.js [hunk 1]
```diff
@@ -141,22 +141,26 @@ const groupData = (nextProps, prevState, nextState) => {
...
-        group.datum[groupBy] = groupValue;
+        group.datum[groupByProperty] = groupValue;
         groups.push(group);
-        groupState[groupValue] = {
-          expanded:
-            prevState.groupState && prevState.groupState[groupValue]
-              ? prevState.groupState[groupValue].expanded
-              : false,
-        };
+        if (prevState.groupState && prevState.groupState[groupValue]) {
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/buildState.js]: I think we should switch this around such that if the caller provides an `expand`, then that takes precedence. If it isn't provided, then the previous state is used. With this approach, a caller can control the expansion state entirely, using `expand` and `onChangeExpand`.  If the caller doesn't want to control it, she can omit `expand`. So, if you want to just set the initial expansion state but then not control it afterwards, you can provide `expand` initially but then omit it afterwards.

Comment (SergiiTykhenkoMCC) [reply src/js/components/DataTable/buildState.js]: Thanks, this part was updated as well as story and doc.js

File: src/js/components/DataTable/buildState.js [hunk 2]
```diff
@@ -152,22 +152,27 @@ const groupData = (nextProps, prevState, nextState) => {
...
-          expanded:
+        if (groupBy.expand) {
+          expandedState = groupBy.expand.some(key => key === groupValue);
+        } else {
+          expandedState =
             prevState.groupState && prevState.groupState[groupValue]
-              ? prevState.groupState[groupValue].expanded
-              : false,
-        };
+              ? prevState.groupState[groupValue]
```
Comment (ericsoderberghp) [inline src/js/components/DataTable/buildState.js]: Should this be `prevState.groupState[groupValue].expanded`? This is illustrated by the "Grouped" story in storybook not working.

**Comment by ericsoderberghp** [comment]
Thanks for fixing the `expanded` non-controlled case. :)

Now, we just need a fresh merge with `master` and update the snapshots, since we updated the dependencies recently, to upgrade jest.

**Comment by SergiiTykhenkoMCC** [comment]
Everything is update

**Comment by ericsoderberghp** [comment]
Thanks for your contribution!

**Pull request merged**